tool/planet/units.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"context"
+
+	"github.com/gravitational/planet/lib/utils"
+
+	systemdDbus "github.com/coreos/go-systemd/dbus"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+func newUnitMonitor(units ...string) (*unitMonitor, error) {
+	conn, err := systemdDbus.NewSystemConnection()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &unitMonitor{
+		conn:  conn,
+		units: units,
+	}, nil
+}
+
+type unitMonitor struct {
+	conn  *systemdDbus.Conn
+	units []string
+}
+
+func (r *unitMonitor) close() {
+	r.conn.Close()
+}
+
+func (r *unitMonitor) start(ctx context.Context) error {
+	resultC := make(chan string, len(r.units))
+	// Number of successfully started start jobs
+	var numStarted int
+	for _, unit := range r.units {
+		logger := log.WithField("unit", unit)
+		logger.Info("Start service.")
+		if _, err := r.conn.StartUnit(unit, "replace", resultC); err != nil {
+			logger.WithError(err).Warn("Failed to start unit.")
+			continue
+		}
+		numStarted += 1
+	}
+	if numStarted == 0 {
+		return nil
+	}
+	r.waitForJobs(ctx, resultC, numStarted)
+	return nil
+}
+
+func (r *unitMonitor) stop(ctx context.Context) error {
+	resultC := make(chan string, len(r.units))
+	// Number of successfully started stop jobs
+	var numStarted int
+	for _, unit := range r.units {
+		logger := log.WithField("unit", unit)
+		logger.Info("Shut down service.")
+		if _, err := r.conn.StopUnit(unit, "replace", resultC); err != nil {
+			logger.WithError(err).Warn("Failed to stop unit.")
+			continue
+		}
+		numStarted += 1
+	}
+	if numStarted != 0 {
+		r.waitForJobs(ctx, resultC, numStarted)
+	}
+	// TODO(dmitri): ListUnitsByPatterns is not supported on systemd 219
+	failedUnits, err := r.conn.ListUnitsFiltered([]string{"failed"})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, unit := range failedUnits {
+		if !utils.StringInSlice(r.units, unit.Name) {
+			continue
+		}
+		if err := r.conn.ResetFailedUnit(unit.Name); err != nil {
+			log.WithError(err).WithField("unit", unit.Name).Warn("Failed to reset failed unit.")
+		}
+	}
+	return nil
+}
+
+func (r *unitMonitor) waitForJobs(ctx context.Context, resultC <-chan string, numStarted int) {
+	var done int
+	for {
+		select {
+		case result := <-resultC:
+			log.WithField("result", result).Info("Job done.")
+			done += 1
+			if done >= numStarted {
```

Comment:
I'm curious whether it's possible that the resultC could miss an event with dbus (ie it get's disconnected, or otherwise lost somehow). If that's happens, trying to trace the code, it looks like waitForJobs could block indefinitely, as the context passed afaict is the context for the overall agent service, and won't timeout until the process is requested to exit.

Reply:
There's a chance if the systemd package or dbus daemon (I guess being the one sending the signal) does not deliver. Otherwise, how can it miss the event (with go-systemd package sending it unconditionally if provided)?

Reply:
Yea, sorry that's sort of what I was getting at, is we're sort of relying on an external behaviour, that there isn't the possibility that it doesn't send or drops events in certain specific conditions.

Reply:
Yes, (potentially) infinite loop is a bit concerning, I think there should be a some sort of timeout here. Maybe pass a `context.WithTimeout()` here.